tutorials/llm/llama/domain-adaptive-pretraining/code/util.py

```diff
@@ -0,0 +1,72 @@
+
+    Args:
+            load_path (str): Path to llama2 weights downlaoded from hugging face
+            save_path (str): Path to save modified dictionary containing the weights.
+
+    Returns:
+            None
+
+    """
+    model_type = "llama2"
+    for i in range(8):
+        print(f"Snapshot {i}")
+        state_dict = torch.load(f"{load_path}/consolidated.0{i}.pth")
+        batch_dict = {}
+        if model_type == "llama2":
+            batch_dict['word_embeddings'] = state_dict['tok_embeddings.weight']
+            batch_dict['output_layer'] = state_dict['output.weight']
+        else:
+            batch_dict['word_embeddings'] = state_dict['model']['embedding.word_embeddings.weight']
+            batch_dict['output_layer'] = state_dict['model']['output_layer.weight']
+        print("Embedding layer dimension: ", batch_dict['word_embeddings'].shape)
+        print("Output layer dimension: ", batch_dict['output_layer'].shape)
+        torch.save(batch_dict, f'{save_path}/embedding_{i}.pt')
+
+
+def merge_embed(old_embd_path, new_embd_path, save_path):
+    "Function to merge embeddings and convert back to hugging face format"
+    model_type = "llama2"
+    for i in range(8):
+        print(f"Snapshot {i}")
+        state_dict = torch.load(f"{old_embd_path}/consolidated.0{i}.pth")
+        batch_dict = torch.load(f'{new_embd_path}/embedding_{i}.pt')
+        if model_type == "llama2":
+            state_dict['output.weight'] = batch_dict['output_layer']
+            state_dict['tok_embeddings.weight'] = batch_dict['word_embeddings']
+            print("embedding shape: ", state_dict['tok_embeddings.weight'].shape)
+            print("output shape: ", state_dict['output.weight'].shape)
+        #             state_dict['args'].padded_vocab_size = state_dict['model']['language_model']['embedding']['word_embeddings']['weight'].shape[0] * 8
+        #             state_dict['args'].vocab_size = state_dict['args'].padded_vocab_size - 768
+        #             print("vocab_size: ", state_dict['args'].vocab_size)
+        #             print("padded_vocab_size: ", state_dict['args'].padded_vocab_size)
+        else:
+            state_dict['tok_embeddings.weight'] = batch_dict['word_embeddings']
+            state_dict['output.weight'] = batch_dict['output_layer']
+            print("embedding shape: ", state_dict['model']['embedding.word_embeddings.weight'].shape)
+            print("output shape: ", state_dict['model']['output_layer.weight'].shape)
+        #             state_dict['args'].padded_vocab_size = state_dict['model']['embedding.word_embeddings.weight'].shape[0] * 8
+        #             state_dict['args'].vocab_size = state_dict['args'].padded_vocab_size - 768
+        #             print("vocab_size: ", state_dict['args'].vocab_size)
+        #             print("padded_vocab_size: ", state_dict['args'].padded_vocab_size)
+        torch.save(state_dict, f"{save_path}/consolidated.0{i}.pth")
```

Comment:
parent dir needs to be created first

Reply:
added a check to ensure that the directory exists before trying to save